sys/linux/vnet.txt

```diff
@@ -738,24 +736,117 @@ ipv6_payload [
+    sack_perm	tcp_sack_perm_option
+    sack		tcp_sack_option
+    timestamp	tcp_timestamp_option
+    md5sig		tcp_md5sig_option
+    fastopen	tcp_fastopen_option
+    exp_fastopen	tcp_exp_fastopen_option
+    exp_smc	tcp_exp_smc_option
+    mptcp       tcp_mptcp_option
+] [varlen]
+
+tcp_mptcp_option [
+    generic     mptcp_generic_option
+    syn         mptcp_syn_option
+    synack      mptcp_synack_option
+    ack         mptcp_ack_option
+    capable     mptcp_capable_option
+    add_addr    mptcp_add_addr_option
+    mp_join     mptcp_mp_join_option
+    mp_flcose   mptcp_mp_fclose_option
+    remove_addr mptcp_remove_addr_option
 ] [varlen]
 
-tcp_option_types = TCPOPT_NOP, TCPOPT_EOL, TCPOPT_MSS, TCPOPT_WINDOW, TCPOPT_SACK_PERM, TCPOPT_SACK, TCPOPT_TIMESTAMP, TCPOPT_MD5SIG, TCPOPT_FASTOPEN, TCPOPT_EXP
+tcp_option_types = TCPOPT_NOP, TCPOPT_EOL, TCPOPT_MSS, TCPOPT_WINDOW, TCPOPT_SACK_PERM, TCPOPT_SACK, TCPOPT_TIMESTAMP, TCPOPT_MD5SIG, TCPOPT_FASTOPEN, TCPOPT_EXP, TCPOPT_MPTCP
+
+mptcp_sub_types = OPTION_TYPE_SYN, OPTION_TYPE_SYNACK, OPTION_TYPE_ACK, OPTION_MP_CAPABLE, OPTION_ADD_ADDR, OPTION_MP_JOIN, OPTION_MP_FCLOSE, OPTION_REMOVE_ADDR
+
+# https://tools.ietf.org/html/rfc6824#section-3
+mptcp_generic_option {
+     type    flags[mptcp_sub_types, int8]
+     length  int8
+     data    array[int8, 0:16]
+ } [packed]
+
+mptcp_capable_option {
+     type        const[TCPOPT_MPTCP, int8]
+     length      int8
+     version     int8:4
+     subtype     const[OPTION_MP_CAPABLE, int8:4]
+     flags       int8
+     sender      int64be
+     receiver    int64be
+ } [packed]
+ 
+ mptcp_add_addr_option {
+     type    const[TCPOPT_MPTCP, int8]
+     length  len[parent, int8]
+     version len[parent, int8:4]
+     subtype const[OPTION_ADD_ADDR, int8:4]
+     port    int16be
+     mac array[int8, 0:8]
+ } [packed]
+ 
+ mptcp_mp_join_option {
+     type    const[TCPOPT_MPTCP, int8]
+     length  int8
+     flags   int8:4
+     subtype const[OPTION_MP_JOIN, int8:4]
+ } [packed]
+ 
+ mptcp_syn_option {
+     type    const[TCPOPT_MPTCP, int8]
+     length  int8
+     flags   int8:4
+     subtype const[OPTION_TYPE_SYN, int8:4]
+     token   int32
+     nonce   int32
+ } [packed]
+
+mptcp_synack_option {
+     type    const[TCPOPT_MPTCP, int8]
+     length  int8
+     flags   int8:4
+     subtype const[OPTION_TYPE_SYNACK, int8:4]
+     mac int64
+     nonce   int32
+ } [packed]
+ 
+ mptcp_ack_option {
+     type    const[TCPOPT_MPTCP, int8]
+     length  int8
+     flags   int8:4
+     subtype const[OPTION_TYPE_ACK, int8:4]
+     mac array[int8, 0:20]
+ } [packed]
+ 
+ mptcp_mp_fclose_option {
+     type    const[TCPOPT_MPTCP, int8]
+     length  int8
+     flags   int16:12
+     subtype const[OPTION_MP_FCLOSE, int16:4]
+     key int64
+ } [packed]
+ 
+ mptcp_remove_addr_option {
+     type    const[TCPOPT_MPTCP, int8]
+     length  int8
+     flags   int8:4
+     subtype const[OPTION_REMOVE_ADDR, int8:4]
+     addrsid int8
```

Comment:
It's possible to provide multiple Address IDs here, so let's make it an `array`.